src/adapter/src/coord/sequencer.rs

```diff
@@ -3510,27 +3446,37 @@ impl<S: Append + 'static> Coordinator<S> {
         session: &Session,
         AlterSourcePlan { id, size, remote }: AlterSourcePlan,
     ) -> Result<ExecuteResponse, AdapterError> {
-        let op = catalog::Op::AlterSource { id, size, remote };
-        self.catalog_transact(Some(session), vec![op]).await?;
-
-        // Re-fetch the updated item from the catalog
-        let entry = self.catalog.get_entry(&id);
-        let full_name = self
+        let source = self
             .catalog
-            .resolve_full_name(entry.name(), Some(session.conn_id()))
-            .to_string();
-        let updated_source = entry.source().ok_or_else(|| CatalogError::UnexpectedType {
-            name: full_name,
-            actual_type: entry.item_type(),
-            expected_type: CatalogItemType::Source,
-        })?;
-        if let DataSourceDesc::Ingestion(Ingestion { host_config, .. }) =
-            &updated_source.data_source
-        {
-            self.controller
-                .storage
-                .alter_collections(vec![(id, host_config.clone())])
-                .await?;
+            .get_entry(&id)
+            .source()
+            .expect("known to be source");
+        match source.data_source {
+            DataSourceDesc::Ingestion(_) => (),
+            DataSourceDesc::Source | DataSourceDesc::Introspection(_) => {
+                coord_bail!("cannot ALTER this type of source");
+            }
+        }
+        let host_config = alter_storage_host_config(size, remote)?;
```

Comment:
`alter_storage_host_config` was hoisted out of the catalog so that its results could be passed to `alter_linked_cluster_ops`.